### ISSUE TITLE
Content repository management

### ISSUE THREAD

**Issue by smashwilson** [deconst/deconst-docs#8]
I'd like to have a web UI to let doc writers easily manage the set of active content repositories. I'm picturing:
1. A writer authenticates to, say, `build.developer.rackspace.com`.
2. They paste the URL of a GitHub repository into a form and click submit.
3. The Strider build is configured automatically, including setting up GitHub webhooks.
4. They're shown a reminder and a link to send a PR to the map repository (#4) to include the content somewhere in the map.

It should also be possible to _remove_ a repository from the site here, too, possibly with a safety guard to prevent removing a repository whose content ID is still referenced by the map.

**Comment by smashwilson** [comment]
Actually, if we wanted to be _extra_ fancy, we could do this automatically with webhooks on the map repository: when new commits are pushed to the map repo, see if any content repositories have been added or removed and adjust the Strider builds appropriately.

Come to think of it, that would be a lot more awesome. What do the rest of you think?

**Comment by smashwilson** [comment]
This is a quick inventory of the webhooks and CI builds that should be automatically configured:
- [x] The **control repository** should have:
  - [x] A webhook that notifies each **mapping service** on push.
  - [x] A webhook that notifies each **layout service** on push.
  - [x] A `.travis.yml` build that runs the mapping and layout services in _validation mode_ to identify any bad syntax or conflicting entries.
  - [x] A `.travis.yml` build that runs the **asset preparer**.
- [ ] Each **content repository** should have:
  - [ ] A `.travis.yml` build that runs the appropriate **preparer**.

The Travis build will become Strider builds once #1 is addressed, but that's not slated for 0.2 and this is.

**Comment by jrperritt** [comment]
> A .travis.yml build that runs the mapping and layout services in validation mode to identify any bad syntax or conflicting entries.

Will we be able to fix the location of the `layout-service` and `mapping-service` in `.travis.yml`? I think this goes back to knowing which services are running on which ports.

**Comment by smashwilson** [comment]
The validation builds shouldn't actually _submit_ content anywhere. They should `pip install -e` the layout and mapping service repos and run them locally against the content being "built." They shouldn't need to know where the services are actually running, then, just where the GitHub repos are.

It would be a pretty trivial thing to set up but the control repository Travis build is already using a Node.js builder for the asset preparer, and I'm not sure how to either activate multiple builds for a single repository, or get a Travis environment with both Python and Node.js installed.

**Comment by jrperritt** [comment]
OK, I think `pip install -e` needs for the repos to be packages (with `setup.py` files) in order to work, so we'll need to make some changes there if we want to do that.

**Comment by smashwilson** [comment]
Oh, right.

We could also be lazy and go the `git clone --depth=1 ${REPO} && python -m whatever` route. Having a proper `setup.py` in place would probably be the better solution, though. :grin:

**Comment by annegentle** [comment]
Does a proper setup.py require that the content repo owners build locally with a Sphinx theme? Or is this setup.py a deconstr script, sort of?

**Comment by smashwilson** [comment]
> Does a proper setup.py require that the content repo owners build locally with a Sphinx theme? Or is this setup.py a deconstr script, sort of?

Oh, these are `setup.py` files for the mapping-service and layout-service repositories themselves (to be able to install them properly as Python packages), not the content repositories.

We do use `conf.py` in content repositories that are Sphinx-based, but the Sphinx preparer will ignore a lot of the layout setup because we're building JSON documents instead.

**Comment by annegentle** [comment]
OK, got it! :100:

**Comment by jrperritt** [comment]
@smashwilson Is there anything more to do for this issue?

**Comment by smashwilson** [comment]
Well - yes, although we should break it out more, and I think it won't be done for 0.2.

This issue is to implement _fully automatic_ setup of hooks and builds on content and control repositories. Meaning, you should be able to turn on a deconst instance configured to point to a given control repository, and automatically get:
1. A webhook installed on the control repository that will send a POST to the webhook services' load balancer.
2. A CI build against the control repository that runs the asset preparer, and the mapping and layout validation logic.
3. A CI build against each content repository that runs the appropriate preparer and submits content to the content services' load balancer. (One tricky part here is finding a way to identify the correct preparer.)

All of these should be safe to run against repositories that are already configured properly (meaning it should do nothing if the configuration is already there). They should handle API key generation and distribution, too. Content repository builds should also be set up for _newly added_ content repositories that are introduced to the content map by regular pull requests.

The specifics of the CI build setup also depends on the way we go for #1.

Basically, this is a giant, nebulous issue that I brain-dumped a long time ago without properly breaking it out :wink: Let's punt this for 0.2 and revisit it later.

**Comment by smashwilson** [comment]
This works now! :sparkles:
